### PR TITLE
Implement ExtensionArray interface

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
First pass to implement the extension array interface. All interface tests are passing except the ones that rely on sorting. 
Still need to better integrate with block code to support released version of pandas as well.

Some other tests are still failing: https://github.com/pandas-dev/pandas/issues/20576, https://github.com/pandas-dev/pandas/issues/20578

### CODE REVIEW

File: geopandas/array.py [hunk 1]
```diff
@@ -362,62 +435,119 @@ def shape(self):
...
         Returns
         -------
-        view : Categorical
-           Returns `self`!
+        uniques : ExtensionArray
         """
-        return self
+        from pandas import factorize
+        _, uniques = factorize(self)
+        return uniques
```
Comment (mrocklin) [inline geopandas/array.py]: Note that two geometries can have the same value but different pointers.  Do we have have a policy on what to do here?

File: geopandas/array.py [hunk 2]
```diff
@@ -362,62 +435,119 @@ def shape(self):
...
 
-    def to_dense(self):
-        """Return my 'dense' representation
+    @property
+    def nbytes(self):
+        return self.data.nbytes
```
Comment (mrocklin) [inline geopandas/array.py]: This is probably an under-estimate.  Geometries are generally far larger than the pointers.  It's possible that GEOS has something for this, although I doubt it.  If it's cheap we might consider converting to WKB and taking the length of that as an approximation.

File: geopandas/array.py [hunk 3]
```diff
@@ -362,62 +435,119 @@ def shape(self):
...
+            An integer ndarray with the factorized values.
+        original : ExtensionArray
+            The original ExtensionArray that factorize was called on.
+
+        See Also
+        --------
+        pandas.factorize
+        ExtensionArray.factorize
+        """
+        return from_wkb(values)
```
Comment (mrocklin) [inline geopandas/array.py]: This implementation seems surprising given the method title.

Comment (mrocklin) [reply geopandas/array.py]: Ah I see, we represent geometries with the appropriate bytestring

Comment (jorisvandenbossche) [reply geopandas/array.py]: Yes, for now I went for using WKB as an intermediate representation for unique / factorize (and possibly also for sorting). With the idea that WKB is the cheapest representation to generate that can be used for algorithms like unique and factorize, without implementing them ourselves (in principle we could also code them in cython using the geos interface for equality and populating a table)

**Comment by jorisvandenbossche** [comment]
Remaining error for pandas master is a bug in extension arrays in pandas: https://github.com/pandas-dev/pandas/issues/20743

**Comment by jorisvandenbossche** [comment]
The ExtensionArray interface how been implemented in the mean time.

**Pull request closed without merging**